client/sftp_client.py

```diff
@@ -56,100 +93,429 @@ def signal_sftp_start(self, _):
+			return task
+		finally:
+			self.not_empty.release()
+
+	def put(self, task):
+		if not isinstance(task, Task):
+			raise TypeError('argument 1 task must be Task instance')
+		with self.not_full:
+			task.register(self.not_empty)
+			self.queue.append(task)
+			self.unfinished_tasks += 1
+			self.not_empty.notify()
+
+	def remove(self, task):
+		with self.mutex:
+			self.queue.remove(task)
+			self.unfinished_tasks += 1
+			self.not_full.notify()
+
+class Task(object):
+	"""
+	Generic task class that contains information about task state
+	and readiness.
+	"""
+	_states = ('Active', 'Cancelled', 'Completed', 'Error', 'Paused', 'Pending')
+	_ready_states = ('Pending',)
+	__slots__ = ('_ready', '_state')
+	def __init__(self, state=None):
+		self._ready = None
+		self._state = None
+		self.state = (state or 'Pending')
+
+	@property
+	def is_done(self):
+		return self._state in ('Cancelled', 'Completed', 'Error')
+
+	@property
+	def is_ready(self):
+		return self._state in self._ready_states
+
+	@property
+	def state(self):
+		return self._state
+
+	@state.setter
+	def state(self, value):
+		if value not in self._states:
+			raise ValueError('invalid state')
+		self._state = value
+		if self._state in self._ready_states and self._ready is not None:
+			self._ready.notify()
+
+	def register(self, ready_event):
+		if self._ready is not None:
+			raise RuntimeError('this task has already been registered')
+		self._ready = ready_event
+
+class ShutdownTask(Task):
+	""" Dummy task used for signaling a shutdown within the queue. """
+	pass
+
+class TransferTask(Task):
+	"""
+	Task used to model transfers. Each task is put in the queue where it will be pass into the
+	_transfer method of the FileManager class for the transfer to occur.
+	"""
+	_states = ('Active', 'Cancelled', 'Completed', 'Error', 'Paused', 'Pending', 'Transferring')
+	__slots__ = ('_state', 'local_path', 'remote_path', 'size', 'transferred', 'treerowref', 'parents', 'init_size', 'show', 'show_children')
+	def __init__(self, local_path, remote_path, parents=None, state=None):
+		super(TransferTask, self).__init__(state=state)
+		self.local_path = local_path
+		self.remote_path = remote_path
+		self.transferred = 0
+		self.size = None
+		self.treerowref = None
+		self.parents = parents
+		self.init_size = None
+		self.show = False
+		self.show_children = False
+
+	@property
+	def progress(self):
+		if self.size is None:
+			percent = 0
+		elif self.size == 0 and self.transferred == 0:
+			percent = 1
+		else:
+			percent = (float(self.transferred) / float(self.size))
+		return min(int(percent * 100), 100)
+
+class DownloadTask(TransferTask):
+	transfer_direction = 'download'
+
+class UploadTask(TransferTask):
+	transfer_direction = 'upload'
+
+class TransferDirectoryTask(TransferTask):
+	"""
+	Task to model a folder transfer. Acts as a parent task
+	to other TransferTasks and is passed into _transfer_folder.
+	"""
+	has_children = False
+	folder_clicked = False
+	dummy_node = None
```

Comment:
This needs a better name, at least indicate the type of what is going to be stored.